internal/credentials/file.go

```diff
@@ -0,0 +1,273 @@
+// Copyright (C) 2024 by Posit Software, PBC.
+
+package credentials
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/pelletier/go-toml/v2"
+	"github.com/posit-dev/publisher/internal/logging"
+	"github.com/posit-dev/publisher/internal/util"
+	"github.com/spf13/afero"
+)
+
+const ondiskFilename = ".connect-credentials"
+
+type fileCredential struct {
+	GUID    string `toml:"guid"`
+	Version uint   `toml:"version"`
+	URL     string `toml:"url"`
+	ApiKey  string `toml:"api_key"`
+}
+
+func (cr *fileCredential) IsValid() bool {
+	return cr.URL != "" && cr.ApiKey != ""
+}
+
+type fileCredentials struct {
+	Credentials map[string]fileCredential `toml:"credentials"`
+}
+
+func (fcs *fileCredentials) CredentialsList() []Credential {
+	var list []Credential
+	for credName, fileCred := range fcs.Credentials {
+		list = append(list, Credential{
+			Name:   credName,
+			GUID:   fileCred.GUID,
+			URL:    fileCred.URL,
+			ApiKey: fileCred.ApiKey,
+		})
+	}
+	return list
+}
+
+func (fcs *fileCredentials) CredentialByGuid(guid string) (Credential, error) {
+	var cred Credential
+	for credName, fileCred := range fcs.Credentials {
+		if fileCred.GUID == guid {
+			cred = Credential{
+				Name:   credName,
+				GUID:   fileCred.GUID,
+				URL:    fileCred.URL,
+				ApiKey: fileCred.ApiKey,
+			}
+			return cred, nil
+		}
+	}
+	return Credential{}, NewNotFoundError(guid)
+}
+
+func (fcs *fileCredentials) RemoveByName(name string) {
+	delete(fcs.Credentials, name)
+}
+
+type fileCredentialsService struct {
+	mu            sync.Mutex
+	afs           afero.Fs
+	log           logging.Logger
+	credsFilepath util.AbsolutePath
+}
+
+func NewFileCredentialsService(log logging.Logger) (*fileCredentialsService, error) {
+	fservice := &fileCredentialsService{
+		log: log,
+	}
+
+	// Set home dir credentials file path
+	homeDir, err := util.UserHomeDir(fservice.afs)
+	if err != nil {
+		return nil, err
+	}
+	fservice.credsFilepath = homeDir.Join(ondiskFilename)
+
+	// Verify file can be modified, will create if not exists
+	err = fservice.setup()
```

Comment:
In general, I prefer the approach that we don't create or validate the file until a request comes in which requires it. That way, the error of it not being able to be accessed is deferred until a requested action, which can then receive the appropriate error. When doing it within the constructor, it is questionable of who is going to see the error and why an empty file is created with no actions.

My requested change is to simply move the `fservice.setup()` method into each of the API request handlers as a first step, returning any errors within those responses.

Please let me know what you think, perhaps I'm missing an advantage for your approach over my recommendation. Thanks!

Reply:
Thanks Bill! This is happening just like you mention, the credentials services constructors run only when endpoints receive requests. Any error that could come up will always be sent and received within an API response.

This PR only has the service introduction, but in the next PR you'll be able to see that such constructors are imported and used only within endpoint handlers :)

Reply:
Thanks Marcos! I think my mind was also raising some concerns about this being a construction method rather than a factory method, so I was not really wanting to return errors from it, but after your explanation and a second look, I thank you very much for your response. I'm good with all of this, I'm approving.

Reply:
Just for clarity to ensure I'm understanding this correctly this will create `.connect-credentials` on machines that do not have keyrings where our extension is loaded and opened since we call `GET /api/credentials` to fetch the Credentials on the machine. Am I correct?

Reply:
Yes, that's right 👍